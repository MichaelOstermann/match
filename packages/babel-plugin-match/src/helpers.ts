import * as t from "@babel/types"
import type { NodePath } from "@babel/core"
import type { LazyExpression, WellFormedObject } from "./types.js"

export class AbortError extends Error {}

export function isAbortError(value: unknown): value is AbortError {
    return value instanceof AbortError
}

export function* findCallExpressions(path: NodePath): Generator<NodePath<t.CallExpression>> {
    let currentPath = path.parentPath
    while (currentPath) {
        if (currentPath.isCallExpression()) {
            yield currentPath
        }
        currentPath = currentPath.parentPath
    }
}

// Plain object with static keys.
export function isWellFormedObject(value: t.Node): value is WellFormedObject {
    if (!t.isObjectExpression(value)) return false
    for (const property of value.properties) {
        if (!t.isObjectProperty(property)) return false
        if (property.computed) return false
        if (!t.isExpression(property.key)) return false
        if (!t.isExpression(property.value)) return false
    }
    return true
}

export function equals(a: t.Expression, b: t.Expression): boolean {
    if (a.type === "Identifier" && b.type === "Identifier") return a.name === b.name
    if (a.type === "StringLiteral" && b.type === "StringLiteral") return a.value === b.value
    if (a.type === "NumericLiteral" && b.type === "NumericLiteral") return a.value === b.value
    return false
}

export function getExpression(lazyExpression: LazyExpression): t.Expression {
    return typeof lazyExpression === "function"
        ? lazyExpression()
        : lazyExpression
}

export function createThrowExpression(value: t.Expression): t.Expression {
    // `Pattern matching error: no pattern matches value ${JSON.stringify(value)}`
    const message = t.templateLiteral([
        t.templateElement({ raw: "Pattern matching error: no pattern matches value " }),
        t.templateElement({ raw: "" }),
    ], [
        t.callExpression(t.memberExpression(t.identifier("JSON"), t.identifier("stringify")), [value]),
    ])

    // (() => { throw new Error(...) })()
    const error = t.newExpression(t.identifier("Error"), [message])
    const throwStatement = t.throwStatement(error)
    return t.callExpression(t.arrowFunctionExpression([], t.blockStatement([
        throwStatement,
    ])), [])
}
