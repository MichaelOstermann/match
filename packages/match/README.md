<div align="center">

<h1>match</h1>

**Zero-runtime exhaustive pattern matching.**

[Documentation](https://MichaelOstermann.github.io/match)

</div>
