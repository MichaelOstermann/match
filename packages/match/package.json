{
    "name": "@monstermann/match",
    "type": "module",
    "version": "0.1.0",
    "description": "Zero-runtime exhaustive pattern matching.",
    "author": "Michael Ostermann <michaelostermann@me.com>",
    "license": "MIT",
    "homepage": "https://MichaelOstermann.github.io/match",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/MichaelOstermann/match.git"
    },
    "bugs": {
        "url": "https://github.com/MichaelOstermann/match/issues"
    },
    "keywords": [
        "pattern",
        "matching",
        "pattern-matching",
        "exhaustive",
        "typescript",
        "match-with",
        "match",
        "switch"
    ],
    "sideEffects": false,
    "exports": {
        ".": {
            "types": "./dist/index.d.ts",
            "import": "./dist/index.js"
        }
    },
    "main": "./dist/index.js",
    "module": "./dist/index.js",
    "types": "./dist/index.d.ts",
    "files": [
        "dist"
    ],
    "scripts": {
        "bundle": "tsdown",
        "release": "bun publish --access public"
    },
    "dependencies": {
        "type-fest": "^4.41.0"
    }
}
