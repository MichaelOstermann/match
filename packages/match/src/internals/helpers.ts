import type { IsLiteral as IsNonNullableLiteral } from "type-fest"

export type IsTrue<T> = true extends T ? true : false
export type IfTrue<T, OnTrue, OnFalse> = true extends T ? OnTrue : OnFalse
export type Union<A, B> = [B] extends [A] ? A : [A] extends [B] ? B : A | B

// `type-fests` definition of literals + null and undefined.
export type IsLiteral<T> = IsNonNullableLiteral<T> extends true
    ? true
    : T extends null | undefined
        ? true
        : false

// @ts-expect-error ignore
export interface MatchError<i> { __: never }
