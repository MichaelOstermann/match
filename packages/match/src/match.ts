import { NonExhaustiveMatch } from "./NonExhaustiveMatch.js"
import type { Match } from "./internals/Match.js"

export function match<const T>(value: T): Match<T> {
    return new NonExhaustiveMatch(value) as unknown as Match<T>
}
