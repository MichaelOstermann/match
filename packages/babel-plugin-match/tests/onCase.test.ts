import { describe, test } from "vitest"
import { expectSnapshot } from "./helpers.js"

describe("onCase", () => {
    test("optimize #1", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, transform)
                .or(false)
        `)
    })

    test("optimize #2", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, () => value)
                .or(false)
        `)
    })

    test("optimize #3", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, (v) => v)
                .or(false)
        `)
    })

    test("optimize #4", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, (v) => v ? 1 : 2)
                .or(false)
        `)
    })

    test("optimize #5", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, (v) => v.foo)
                .or(false)
        `)
    })

    test("optimize #6", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, (v) => transform(v))
                .or(false)
        `)
    })

    test("optimize #7", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(value)
                .onCase(a, (v) => transform(v.foo))
                .or(false)
        `)
    })

    test("optimize #8", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, transform)
                .or(false)
        `)
    })

    test("optimize #9", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, () => value)
                .or(false)
        `)
    })

    test("optimize #10", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, (v) => v)
                .or(false)
        `)
    })

    test("optimize #11", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, (v) => v ? 1 : 2)
                .or(false)
        `)
    })

    test("optimize #12", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, (v) => v.foo)
                .or(false)
        `)
    })

    test("optimize #13", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, (v) => transform(v))
                .or(false)
        `)
    })

    test("optimize #14", () => {
        expectSnapshot(`
            import { match } from "@monstermann/match"
            match(getValue())
                .onCase(a, (v) => transform(v.foo))
                .or(false)
        `)
    })
})
