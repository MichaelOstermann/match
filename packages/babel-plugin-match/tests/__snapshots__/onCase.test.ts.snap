// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`onCase > optimize #1 1`] = `
"import { match } from "@monstermann/match";
value === a ? transform(value) : false;"
`;

exports[`onCase > optimize #2 1`] = `
"import { match } from "@monstermann/match";
value === a ? value : false;"
`;

exports[`onCase > optimize #3 1`] = `
"import { match } from "@monstermann/match";
value === a ? v : false;"
`;

exports[`onCase > optimize #4 1`] = `
"import { match } from "@monstermann/match";
value === a ? value ? 1 : 2 : false;"
`;

exports[`onCase > optimize #5 1`] = `
"import { match } from "@monstermann/match";
value === a ? value.foo : false;"
`;

exports[`onCase > optimize #6 1`] = `
"import { match } from "@monstermann/match";
value === a ? transform(value) : false;"
`;

exports[`onCase > optimize #7 1`] = `
"import { match } from "@monstermann/match";
value === a ? transform(value.foo) : false;"
`;

exports[`onCase > optimize #8 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? transform(_value) : false)(getValue());"
`;

exports[`onCase > optimize #9 1`] = `
"import { match } from "@monstermann/match";
getValue() === a ? value : false;"
`;

exports[`onCase > optimize #10 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? v : false)(getValue());"
`;

exports[`onCase > optimize #11 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? _value ? 1 : 2 : false)(getValue());"
`;

exports[`onCase > optimize #12 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? _value.foo : false)(getValue());"
`;

exports[`onCase > optimize #13 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? transform(_value) : false)(getValue());"
`;

exports[`onCase > optimize #14 1`] = `
"import { match } from "@monstermann/match";
(_value => _value === a ? transform(_value.foo) : false)(getValue());"
`;
